models/classes/class.QtiTestService.php

```diff
@@ -866,24 +869,62 @@ protected function importTest(
         return $report;
     }
 
+    /**
+     * @throws common_Exception
+     */
     private function deleteTestsFromClassByLabel(
         string $testLabel,
         string $itemsClassLabel,
         core_kernel_classes_Resource $testClass,
         core_kernel_classes_Class $itemClass
     ): void {
-        $testService = $this->getTestService();
-        $itemTreeService = $this->getItemTreeService();
+        $itemClassesToDelete = $this->getSubclassesByLabel($itemClass, $itemsClassLabel);
+
+        /** @var string[] $refs */
+        $refs = $this->collectResourceReferences($itemClassesToDelete);
 
+        /** @var string[] $deletedTestsUris */
+        $deletedTestsUris = [];
         foreach ($testClass->getInstances() as $testInstance) {
+            /** @var core_kernel_classes_Resource $testInstance */
             if ($testInstance->getLabel() === $testLabel) {
-                $testService->deleteTest($testInstance);
+                $deletedTestsUris[] = $testInstance->getUri();
             }
         }
 
-        foreach ($itemClass->getSubClasses() as $subClass) {
-            if ($subClass->getLabel() === $itemsClassLabel) {
-                $itemTreeService->deleteClass($subClass);
+        /** @var string[] $deletedItemClassesUris */
+        $deletedItemClassesUris = [];
+        foreach ($itemClassesToDelete as $subClass) {
+            $deletedItemClassesUris[] = $subClass->getUri();
+        }
+
+        // Trigger the deletion event so extensions holding references to
+        // the deleted test or items can remove them first.
+        //
```

Comment:
I do not recommend to add this type of commenting here. You are basically explaining why you took the decision of to call this method. This should be corrected by renaming the event class to something like `DeleteAssetsEvent` as I mentioned before. Anybody reading this will know this event is deleting assets :)

Reply:
Thanks!